### PR TITLE
Add Implicit Quantile Network

### LINKED ISSUE #370: Implicit Quantile Network (IQN) and Fully parameterized Quantile Function (FQF)
- [x] I have marked all applicable categories:
    + [ ] exception-raising bug
    + [ ] RL algorithm bug
    + [ ] documentation request (i.e. "X is missing from the documentation.")
    + [x] new feature request
- [x] I have visited the [source website](https://github.com/thu-ml/tianshou/)
- [x] I have searched through the [issue tracker](https://github.com/thu-ml/tianshou/issues) for duplicates
- [ ] I have mentioned version numbers, operating system and environment, where applicable:
  ```python
  import tianshou, torch, numpy, sys
  print(tianshou.__version__, torch.__version__, numpy.__version__, sys.version, sys.platform)
  ```

I would like to take a stab at implementing [IQN](https://arxiv.org/pdf/1806.06923.pdf) and [FQF](https://arxiv.org/pdf/1911.02140.pdf), two distributional RL model with great results on Atari games, following the excellent reference implementations [here](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch).

I am almost done with IQN and will submit a PR soon. Before I move on to FQF, I'd highly appreciate your feedback.

### PR DESCRIPTION
**Pull request by nuance1979**
- Add Implicit Quantile Network
- Reference: [paper](https://arxiv.org/pdf/1806.06923.pdf) and [code](https://github.com/ku2482/fqf-iqn-qrdqn.pytorch)

I will try to get some results over the weekend but our GPU machines are a bit busy recently.

### CODE REVIEW

File: tianshou/policy/modelfree/iqn.py [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    """Implementation of Implicit Quantile Network. arXiv:1806.06923.
+
+    :param torch.nn.Module model: a model following the rules in
+        :class:`~tianshou.policy.BasePolicy`. (s -> logits)
+    :param torch.optim.Optimizer optim: a torch.optim for optimizing the model.
+    :param float discount_factor: in [0, 1].
+    :param int K: the number of samples for policy evaluation. Default to 32.
+    :param int N: the number of samples for online model in training. Default to 8.
+    :param int N_prime: the number of samples for target model in training.
+        Default to 8.
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/iqn.py]: Do you have other better names for these variables?

Comment (nuance1979) [reply tianshou/policy/modelfree/iqn.py]: Good point. Done.

File: tianshou/utils/net/discrete.py [hunk 2]
```diff
@@ -111,3 +111,106 @@ def forward(
...
+        # Calculate cos(i * \pi * \tau).
+        cosines = torch.cos(taus.view(batch_size, N, 1) * i_pi).view(
+            batch_size * N, self.num_cosines
+        )
+        # Calculate embeddings of taus.
+        tau_embeddings = self.net(cosines).view(batch_size, N, self.embedding_dim)
+        return tau_embeddings
+
+
+class IQN(Critic):
```
Comment (Trinkle23897) [inline tianshou/utils/net/discrete.py]: I personally recommend moving this network under test_iqn.py because few algorithms reuse this network architecture. But anyway it's just my personal thought, it's okay if you insist.

Comment (nuance1979) [reply tianshou/utils/net/discrete.py]: I agree that it may not be "common" enough to put it here but I feel putting it in `test_iqn.py` is not appropriate, either, because I need it in `atari_iqn.py` and importing from a test file is very weird. The other way around (`test_iqn.py` importing this class from `atari_iqn.py`) is also awkward. Therefore, before we find a more suitable place for this class, I'd rather let it be here.

I am not sure about the naming. Do you think it'd be better if I change it from `IQN` to `ImplicitQuantileNetwork`?

File: examples/atari/README.md [hunk 3]
```diff
@@ -54,6 +54,20 @@ One epoch here is equal to 100,000 env step, 100 epochs stand for 10M.
 | SeaquestNoFrameskip-v4      | 5676        | ![](results/qrdqn/Seaquest_rew.png)     | `python3 atari_qrdqn.py --task "SeaquestNoFrameskip-v4"`  |
 | SpaceInvadersNoFrameskip-v4      | 938        | ![](results/qrdqn/SpaceInvader_rew.png)     | `python3 atari_qrdqn.py --task "SpaceInvadersNoFrameskip-v4"`  |
 
+# IQN (single run)
+
+One epoch here is equal to 100,000 env step, 100 epochs stand for 10M.
+
+| task                        | best reward | reward curve                          | parameters                                                   |
+| --------------------------- | ----------- | ------------------------------------- | ------------------------------------------------------------ |
+| PongNoFrameskip-v4          | 20.2        | ![](results/iqn/Pong_rew.png)         | `python3 atari_iqn.py --task "PongNoFrameskip-v4" --batch-size 64` |
```
Comment (nuance1979) [inline examples/atari/README.md]: @Trinkle23897 I assume that the "best reward" is from `grep best_reward log`, i.e., the best reward during training, not the mean reward during testing. Correct?

File: examples/atari/atari_iqn.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+    parser.add_argument('--eps-train', type=float, default=1.)
+    parser.add_argument('--eps-train-final', type=float, default=0.05)
+    parser.add_argument('--buffer-size', type=int, default=100000)
+    parser.add_argument('--lr', type=float, default=0.0001)
+    parser.add_argument('--gamma', type=float, default=0.99)
+    parser.add_argument('--sample-size', type=int, default=32)
+    parser.add_argument('--online-sample-size', type=int, default=8)
+    parser.add_argument('--target-sample-size', type=int, default=8)
+    parser.add_argument('--num-cosines', type=int, default=64)
+    parser.add_argument('--hidden-sizes', type=int, nargs='*', default=[512, 512])
```
Comment (Trinkle23897) [inline examples/atari/atari_iqn.py]: why did you change this?

File: examples/atari/README.md (hunk 3 above)
Comment (Trinkle23897) [reply examples/atari/README.md]: It's the best reward over each test result in each epoch test phase

File: examples/atari/atari_iqn.py (hunk 4 above)
Comment (nuance1979) [reply examples/atari/atari_iqn.py]: This is to make the network structure comparable to QRDQN. This `hidden-sizes` parameter controls the MLP structure **after** the feature net. See https://github.com/thu-ml/tianshou/blob/458028a326fed7bffacd42b457a9a55e933a7bd3/examples/atari/atari_network.py#L34-L37

Comment (Trinkle23897) [reply examples/atari/atari_iqn.py]: Yes but this equals to `[512]` instead of `[512, 512]`.

Comment (nuance1979) [reply examples/atari/atari_iqn.py]: Ah, you are right. I'll change it back and rerun the experiments.

File: tianshou/policy/modelfree/iqn.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+        super().__init__(model, optim, discount_factor, sample_size, estimation_step,
+                         target_update_freq, reward_normalization, **kwargs)
+        assert sample_size > 1, "sample_size should be greater than 1"
+        assert online_sample_size > 1, "online_sample_size should be greater than 1"
+        assert target_sample_size > 1, "target_sample_size should be greater than 1"
+        self._sample_size = sample_size  # for policy eval
+        self._online_sample_size = online_sample_size
+        self._target_sample_size = target_sample_size
+        # set sample size for online and target model
+        self.model.sample_size = self._online_sample_size  # type: ignore
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/iqn.py]: plz do not do this magic
is it possible to add an extra argument in net.forward?

Comment (nuance1979) [reply tianshou/policy/modelfree/iqn.py]: Good point. I did consider adding it as an extra argument in `forward()` but somehow thought it was more complicated. Now I look at it again and it seems I should put `sample_size` completely in the model, instead of in the policy. I'll make the change.

File: tianshou/utils/net/discrete.py [hunk 6]
```diff
@@ -111,3 +111,94 @@ def forward(
...
+        super().__init__(preprocess_net, hidden_sizes, last_size,
+                         preprocess_net_output_dim, device)
+        self.input_dim = getattr(preprocess_net, "output_dim",
+                                 preprocess_net_output_dim)
+        self.embed_model = CosineEmbeddingNetwork(num_cosines,
+                                                  self.input_dim).to(device)
+
+    def forward(  # type: ignore
+        self, s: Union[np.ndarray, torch.Tensor], sample_size: int, **kwargs: Any
+    ) -> Tuple[torch.Tensor, Any]:
```
Comment (Trinkle23897) [inline tianshou/utils/net/discrete.py]: hmm does this change the definition of network?
https://tianshou.readthedocs.io/en/master/tutorials/dqn.html#build-the-network
The first return object can be any form but the second one should be the hidden state

File: tianshou/policy/modelfree/iqn.py [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        batch: Batch,
+        state: Optional[Union[dict, Batch, np.ndarray]] = None,
+        model: str = "model",
+        input: str = "obs",
+        **kwargs: Any,
+    ) -> Batch:
+        model = getattr(self, model)
+        obs = batch[input]
+        obs_ = obs.obs if hasattr(obs, "obs") else obs
+        if model == "model_old":
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/iqn.py:65]: this is a bug. model now is a `nn.Module` instead of str

File: tianshou/policy/modelfree/iqn.py [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+            sample_size = self._target_sample_size
+        elif self.training:
+            sample_size = self._online_sample_size
+        else:
+            sample_size = self._sample_size
+        logits, h = model(obs_, sample_size, state=state, info=batch.info)
```
Comment (Trinkle23897) [inline tianshou/policy/modelfree/iqn.py]: better using `sample_size=sample_size`

File: tianshou/utils/net/discrete.py (hunk 6 above)
Comment (nuance1979) [reply tianshou/utils/net/discrete.py]: I see. I'll follow the convention.

File: tianshou/policy/modelfree/iqn.py (hunk 7 above)
Comment (nuance1979) [reply tianshou/policy/modelfree/iqn.py:65]: Good catch! Thanks!

**Review by Trinkle23897 — APPROVED** [review]
